crates/turborepo-lib/src/shim.rs

```diff
@@ -227,114 +226,156 @@ pub struct RepoState {
+        match &self.workspace_globs {
+            Some(globs) => globs
+                .test(self.path.to_path_buf(), target_path)
+                .unwrap_or(false),
+            None => false,
+        }
+    }
+}
 
-            let pnpm = PackageManager::Pnpm;
-            let npm = PackageManager::Npm;
-            let is_workspace =
-                pnpm.get_workspace_globs(dir).is_ok() || npm.get_workspace_globs(dir).is_ok();
+impl RepoState {
+    fn generate_potential_turbo_roots(reference_dir: &Path) -> Vec<InferInfo> {
+        // Find all directories that contain a `package.json` or a `turbo.json`.
+        // Gather a bit of additional metadata about them.
+        let potential_turbo_roots: Vec<_> = reference_dir
+            .ancestors()
+            .map(|path| {
+                let workspace_globs = PackageManager::Pnpm
+                    .get_workspace_globs(path)
+                    .unwrap_or_else(|_| {
+                        PackageManager::Npm
+                            .get_workspace_globs(path)
+                            .unwrap_or(None)
+                    });
+
+                InferInfo {
+                    path: path.to_owned(),
+                    has_package_json: fs::metadata(path.join("package.json")).is_ok(),
+                    has_turbo_json: fs::metadata(path.join("turbo.json")).is_ok(),
+                    package_has_workspaces: workspace_globs.is_some(),
+                    workspace_globs,
+                }
+            })
+            .filter(|info| info.has_package_json || info.has_turbo_json)
+            .collect();
 
-            if is_workspace {
-                let local_turbo_state = LocalTurboState::infer(dir);
+        potential_turbo_roots
+    }
 
+    fn process_potential_turbo_roots(potential_turbo_roots: Vec<InferInfo>) -> Result<Self> {
+        // Potential improvements:
+        // - Detect invalid configuration where turbo.json isn't peer to package.json.
+        // - There are a couple of possible early exits to prevent traversing all the
+        //   way to root at significant code complexity increase.
+        //
+        //   1. [0].has_turbo_json && [0].package_has_workspaces
+        //   2. [0].has_turbo_json && [n].has_turbo_json && [n].is_workspace_root_of(0)
+        //
+        //   If we elect to make any of the changes for early exits we need to expand
+        // the   test suite which presently relies on the fact that the
+        // selection runs in a   loop to avoid creating those test cases.
+
+        // We need to perform the same search strategy for _both_ turbo.json and _then_
+        // package.json.
+        let search_locations = [InferInfo::has_turbo_json, InferInfo::has_package_json];
+
+        for check_set_comparator in search_locations {
+            let check_roots: Vec<&InferInfo> = potential_turbo_roots
+                .iter()
+                .filter(check_set_comparator)
+                .collect();
```

Comment:
More of a note than a serious suggestion, but we could probably do this without using `collect` (and therefore without allocating a Vec) if we use `peek`. We'd call `next` and get the first element.

Reply:
The test coverage actually made this really easy to do at high confidence, so I did it. 👍